File: domain/src/main/java/com/github/anrimian/musicplayer/domain/interactors/library/LibraryFoldersScreenInteractor.java
package com.github.anrimian.musicplayer.domain.interactors.library;

import com.github.anrimian.musicplayer.domain.models.composition.Composition;
import com.github.anrimian.musicplayer.domain.models.folders.FileSource;
import com.github.anrimian.musicplayer.domain.models.folders.FolderFileSource;
import com.github.anrimian.musicplayer.domain.models.folders.IgnoredFolder;
import com.github.anrimian.musicplayer.domain.models.order.Order;
import com.github.anrimian.musicplayer.domain.models.playlist.PlayList;
import com.github.anrimian.musicplayer.domain.repositories.EditorRepository;
import com.github.anrimian.musicplayer.domain.repositories.LibraryRepository;
import com.github.anrimian.musicplayer.domain.repositories.MediaScannerRepository;

import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.List;

import javax.annotation.Nullable;

import io.reactivex.Completable;
import io.reactivex.Observable;
import io.reactivex.Single;
import io.reactivex.subjects.BehaviorSubject;

/**
 * Created on 24.10.2017.
 */

public class LibraryFoldersScreenInteractor {

    private final LibraryFoldersInteractor foldersInteractor;
    private final LibraryRepository libraryRepository;
    private final EditorRepository editorRepository;
    private final MediaScannerRepository mediaScannerRepository;


    private final BehaviorSubject<Boolean> moveModeSubject = BehaviorSubject.createDefault(false);
    private final LinkedHashSet<FileSource> filesToCopy = new LinkedHashSet<>();
    private final LinkedHashSet<FileSource> filesToMove = new LinkedHashSet<>();

    @Nullable
    private Long moveFromFolderId;

    public LibraryFoldersScreenInteractor(LibraryFoldersInteractor foldersInteractor,
                                          LibraryRepository libraryRepository,
                                          EditorRepository editorRepository,
                                          MediaScannerRepository mediaScannerRepository) {
        this.foldersInteractor = foldersInteractor;
        this.libraryRepository = libraryRepository;
        this.editorRepository = editorRepository;
        this.mediaScannerRepository = mediaScannerRepository;
    }

    public Observable<List<FileSource>> getFoldersInFolder(@Nullable Long folderId,
                                                            @Nullable String searchQuery) {
        return foldersInteractor.getFoldersInFolder(folderId, searchQuery);
    }

    public Observable<FolderFileSource> getFolderObservable(long folderId) {
        return foldersInteractor.getFolderObservable(folderId);
    }

    public void playAllMusicInFolder(@Nullable Long folderId) {
        foldersInteractor.playAllMusicInFolder(folderId);
    }

    public Single<List<Composition>> getAllCompositionsInFolder(@Nullable Long folderId) {
        return foldersInteractor.getAllCompositionsInFolder(folderId);
    }

    public Single<List<Composition>> getAllCompositionsInFileSources(List<FileSource> fileSources) {
        return foldersInteractor.getAllCompositionsInFileSources(fileSources);
    }

    public void play(List<FileSource> fileSources) {
        foldersInteractor.play(fileSources);
    }

    public void play(Long folderId, Composition composition) {
        foldersInteractor.play(folderId, composition);
    }

    public Single<List<Composition>> addCompositionsToPlayNext(List<FileSource> fileSources) {
        return foldersInteractor.addCompositionsToPlayNext(fileSources);
    }

    public Single<List<Composition>> addCompositionsToEnd(List<FileSource> fileSources) {
        return foldersInteractor.addCompositionsToEnd(fileSources);
    }

    public Single<List<Composition>> deleteFiles(List<FileSource> fileSources) {
        return foldersInteractor.deleteCompositions(fileSources);
    }

    public Single<List<Composition>> deleteFolder(FolderFileSource folder) {
        return foldersInteractor.deleteFolder(folder);
    }

    public Single<List<Composition>> addCompositionsToPlayList(Long folderId, PlayList playList) {
        return foldersInteractor.addCompositionsToPlayList(folderId, playList);
    }

    public Single<List<Composition>> addCompositionsToPlayList(List<FileSource> fileSources, PlayList playList) {
        return foldersInteractor.addCompositionsToPlayList(fileSources, playList);
    }

    public void setFolderOrder(Order order) {
        foldersInteractor.setFolderOrder(order);
    }

    public Order getFolderOrder() {
        return foldersInteractor.getFolderOrder();
    }

    public void saveCurrentFolder(@Nullable Long folderId) {
        foldersInteractor.saveCurrentFolder(folderId);
    }

    public Single<List<Long>> getCurrentFolderScreens() {
        return foldersInteractor.getCurrentFolderScreens();
    }

    public Completable renameFolder(long folderId, String newName) {
        return foldersInteractor.renameFolder(folderId, newName);
    }

    public void addFilesToMove(@Nullable Long folderId, Collection<FileSource> fileSources) {
        filesToMove.clear();
        filesToMove.addAll(fileSources);
        this.moveFromFolderId = folderId;
        moveModeSubject.onNext(true);
    }

    public void addFilesToCopy(@Nullable Long folderId, Collection<FileSource> fileSources) {
        filesToCopy.clear();
        filesToCopy.addAll(fileSources);
        this.moveFromFolderId = folderId;
        moveModeSubject.onNext(true);
    }

    public void stopMoveMode() {
        filesToCopy.clear();
        filesToMove.clear();
        moveFromFolderId = null;
        moveModeSubject.onNext(false);
    }

    public Completable moveFilesTo(@Nullable Long folderId) {
        Completable completable;
        if (!filesToMove.isEmpty()) {
            completable = editorRepository.moveFiles(filesToMove, moveFromFolderId, folderId);
        } else if (!filesToCopy.isEmpty()) {
            completable = Completable.error(new Exception("not implemented"));
        } else {
            throw new IllegalStateException("unexpected state");
        }
        return completable.doOnComplete(this::stopMoveMode);
    }

    public Completable moveFilesToNewFolder(@Nullable Long folderId, String folderName) {
        Completable completable;
        if (!filesToMove.isEmpty()) {
            completable = editorRepository.moveFilesToNewDirectory(filesToMove,
                    moveFromFolderId,
                    folderId,
                    folderName);
        } else if (!filesToCopy.isEmpty()) {
            completable = Completable.error(new Exception("not implemented"));
        } else {
            throw new IllegalStateException("unexpected state");
        }
        return completable.doOnComplete(this::stopMoveMode);
    }

    public BehaviorSubject<Boolean> getMoveModeObservable() {
        return moveModeSubject;
    }

    public LinkedHashSet<FileSource> getFilesToMove() {
        return filesToMove;
    }

    public Single<IgnoredFolder> addFolderToIgnore(FolderFileSource folder) {
        return foldersInteractor.addFolderToIgnore(folder);
    }

    public Completable deleteIgnoredFolder(IgnoredFolder folder) {
        return libraryRepository.deleteIgnoredFolder(folder)
                .andThen(mediaScannerRepository.runStorageScanner());
    }
}

File: domain/src/main/java/com/github/anrimian/musicplayer/domain/interactors/library/LibraryFoldersScreenInteractor.kt
package com.github.anrimian.musicplayer.domain.interactors.library

import com.github.anrimian.musicplayer.domain.Constants
import com.github.anrimian.musicplayer.domain.models.composition.Composition
import com.github.anrimian.musicplayer.domain.models.composition.DeletedComposition
import com.github.anrimian.musicplayer.domain.models.folders.FileSource
import com.github.anrimian.musicplayer.domain.models.folders.FolderFileSource
import com.github.anrimian.musicplayer.domain.models.folders.FolderInfo
import com.github.anrimian.musicplayer.domain.models.folders.IgnoredFolder
import com.github.anrimian.musicplayer.domain.models.order.Order
import com.github.anrimian.musicplayer.domain.models.utils.ListPosition
import com.github.anrimian.musicplayer.domain.repositories.LibraryRepository
import com.github.anrimian.musicplayer.domain.repositories.UiStateRepository
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.subjects.BehaviorSubject

/**
 * Created on 24.10.2017.
 */
class LibraryFoldersScreenInteractor(
    private val foldersInteractor: LibraryFoldersInteractor,
    private val libraryRepository: LibraryRepository,
    private val uiStateRepository: UiStateRepository
) {

    private val moveModeSubject = BehaviorSubject.createDefault(false)

    private val filesToCopy: LinkedHashSet<FileSource> = LinkedHashSet()
    private val filesToMove: LinkedHashSet<FileSource> = LinkedHashSet()

    private var moveFromFolderId: Long? = null

    fun getFoldersInFolder(
        folderId: Long?,
        searchQuery: String?
    ): Observable<List<FileSource>> {
        //on receive remove files from move/copy of they are not present in list
        return foldersInteractor.getFoldersInFolder(folderId, searchQuery)
    }

    fun getFolderObservable(folderId: Long): Observable<FolderInfo> {
        return foldersInteractor.getFolderObservable(folderId)
    }

    fun playAllMusicInFolder(folderId: Long?): Completable {
        return foldersInteractor.playAllMusicInFolder(folderId)
    }

    fun getAllCompositionsInFolder(folderId: Long?): Single<List<Composition>> {
        return foldersInteractor.getAllCompositionsInFolder(folderId)
    }

    fun getAllCompositionsInFileSources(fileSources: List<FileSource>): Single<List<Composition>> {
        return foldersInteractor.getAllCompositionsInFileSources(fileSources)
    }

    fun play(fileSources: Collection<FileSource>, position: Int = Constants.NO_POSITION): Completable{
        return foldersInteractor.play(ArrayList(fileSources), position)
    }

    fun deleteFiles(fileSources: List<FileSource>): Single<List<DeletedComposition>> {
        return foldersInteractor.deleteFiles(fileSources)
    }

    fun deleteFolder(folder: FolderFileSource): Single<List<DeletedComposition>> {
        return foldersInteractor.deleteFolder(folder)
    }

    fun setFolderOrder(order: Order) {
        foldersInteractor.setFolderOrder(order)
    }

    fun getFolderOrder(): Order {
        return foldersInteractor.getFolderOrder()
    }

    fun saveCurrentFolder(folderId: Long?) {
        foldersInteractor.saveCurrentFolder(folderId)
    }

    fun getCurrentFolderScreens(): Single<List<Long>> {
        return foldersInteractor.getCurrentFolderScreens()
    }

    fun getParentFolders(compositionId: Long): Single<List<Long>> {
        return foldersInteractor.getParentFolders(compositionId)
    }

    fun renameFolder(folderId: Long, newName: String): Completable {
        return foldersInteractor.renameFolder(folderId, newName)
    }

    fun addFilesToMove(folderId: Long?, fileSources: Collection<FileSource>) {
        filesToMove.clear()
        filesToMove.addAll(fileSources)
        moveFromFolderId = folderId
        moveModeSubject.onNext(true)
    }

    fun addFilesToCopy(folderId: Long?, fileSources: Collection<FileSource>) {
        filesToCopy.clear()
        filesToCopy.addAll(fileSources)
        moveFromFolderId = folderId
        moveModeSubject.onNext(true)
    }

    fun stopMoveMode() {
        filesToCopy.clear()
        filesToMove.clear()
        moveFromFolderId = null
        moveModeSubject.onNext(false)
    }

    fun moveFilesTo(folderId: Long?): Completable {
        val completable = if (filesToMove.isNotEmpty()) {
            foldersInteractor.moveFiles(filesToMove, moveFromFolderId, folderId)
        } else if (filesToCopy.isNotEmpty()) {
            Completable.error(Exception("not implemented"))
        } else {
            Completable.complete()
        }
        return completable.doOnComplete { stopMoveMode() }
    }

    fun moveFilesToNewFolder(folderId: Long?, folderName: String): Completable {
        val completable = if (filesToMove.isNotEmpty()) {
            foldersInteractor.moveFilesToNewDirectory(
                filesToMove,
                moveFromFolderId,
                folderId,
                folderName
            )
        } else if (filesToCopy.isNotEmpty()) {
            Completable.error(Exception("not implemented"))
        } else {
            Completable.complete()
        }
        return completable.doOnComplete { stopMoveMode() }
    }

    fun getMoveModeObservable(): BehaviorSubject<Boolean> {
        return moveModeSubject
    }

    fun getFilesToMove(): LinkedHashSet<FileSource> {
        return filesToMove
    }

    fun addFolderToIgnore(folder: FolderFileSource): Single<IgnoredFolder> {
        return foldersInteractor.addFolderToIgnore(folder)
    }

    fun deleteIgnoredFolder(folder: IgnoredFolder): Completable {
        return foldersInteractor.deleteIgnoredFolder(folder)
    }

    fun saveListPosition(folderId: Long?, listPosition: ListPosition) {
        uiStateRepository.saveFolderListPosition(folderId, listPosition)
    }

    fun getSavedListPosition(folderId: Long?): ListPosition? {
        return uiStateRepository.getSavedFolderListPosition(folderId)
    }
}